lib/backend/k8s/k8s_test.go

```diff
@@ -2304,6 +2326,152 @@ var _ = testutils.E2eDatastoreDescribe("Test Watch support", testutils.Datastore
+			var kvp1or2 *model.KVPair
+			for _, kvp := range l.KVPairs {
+				k := kvp.Key.(model.ResourceKey)
+				if k.Name == "test-net-policy-1" || k.Name == "test-net-policy-2" {
+					policy := kvp.Value.(*apiv3.NetworkPolicy)
+					policy.SetLabels(map[string]string{"test": "00"})
+					kvp1or2, err = c.Update(ctx, kvp)
+					Expect(err).ToNot(HaveOccurred())
+					found++
+				}
+			}
+			Expect(found).To(Equal(2))
+
+			// Modify the kubernetes policies
+			for _, name := range []string{"test-net-policy-3", "test-net-policy-4"} {
+				p, err := c.ClientSet.NetworkingV1().NetworkPolicies("default").Get(name, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				p.SetLabels(map[string]string{"test": "00"})
+				_, err = c.ClientSet.NetworkingV1().NetworkPolicies("default").Update(p)
+				Expect(err).ToNot(HaveOccurred())
+			}
+
+			log.WithField("revision", l.Revision).Info("[TEST] first watch")
+			watch, err := c.Watch(ctx, model.ResourceListOptions{Kind: apiv3.KindNetworkPolicy}, l.Revision)
+			Expect(err).NotTo(HaveOccurred())
+
+			// We should see 4 events
+			event := ExpectModifiedEvent(watch.ResultChan())
+			log.WithField("revision", event.New.Revision).Info("[TEST] first event")
+			event = ExpectModifiedEvent(watch.ResultChan())
+			log.WithField("revision", event.New.Revision).Info("[TEST] second event")
+			event = ExpectModifiedEvent(watch.ResultChan())
+			log.WithField("revision", event.New.Revision).Info("[TEST] third event")
+			event = ExpectModifiedEvent(watch.ResultChan())
+			log.WithField("revision", event.New.Revision).Info("[TEST] fourth event")
+
+			// There should be no more events
+			Expect(watch.ResultChan()).ToNot(Receive())
+
+			watch.Stop()
+
+			// Make a second change
+			kvp1or2.Value.(*apiv3.NetworkPolicy).SetLabels(map[string]string{"test":"01"})
+			_, err = c.Update(ctx, kvp1or2)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Resume watching at the revision of the event we got
+			log.WithField("revision", event.New.Revision).Info("second watch")
+			watch, err = c.Watch(ctx, model.ResourceListOptions{Kind: apiv3.KindNetworkPolicy}, event.New.Revision)
+			Expect(err).NotTo(HaveOccurred())
+
+			// We should only get 1 update, because the event from the previous watch should have been "latest"
+			ExpectModifiedEvent(watch.ResultChan())
+
+			// There should be no more events
+			Expect(watch.ResultChan()).ToNot(Receive())
+
+			watch.Stop()
+		})
+
+		It("supports watching from part way through a list", func() {
+			l, err := c.List(ctx, model.ResourceListOptions{Kind:apiv3.KindNetworkPolicy}, "")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(l.KVPairs).To(HaveLen(4))
+
+			// Watch from part way
+			for i := 0; i<4; i++ {
+				revision := l.KVPairs[i].Revision
+				log.WithFields(log.Fields{
+					"revision": revision,
+					"key": l.KVPairs[i].Key.String()}).Info("[Test] starting watch")
+				_, err := c.Watch(ctx, model.ResourceListOptions{Kind: apiv3.KindNetworkPolicy}, revision)
```

Comment:
Do we want to stop the watch here as well?

Reply:
Good catch.